### ISSUE TITLE
Declaring boost dependency with version requirement does not work properly 

### ISSUE THREAD

**Issue by edvardsp** [mesonbuild/meson#1475]
I have declared a boost dependency with a version constraint as such:

```
boost_dep = dependency('boost',
    modules : ['context'],
    version : '>=1.61',
    required : true)
```

On my computer I've got boost::context version 1.58 installed. When invoking ninja, the following is printed:

```
Dependency Boost (context) found: YES 1.58
```

I expected dependency to fail as the version constraint is not satisfied.

Doing a direct version_compare of the returned dependency object seems to work as intended:

```
message(boost_dep.version().version_compare('>=1.61'))
  => (when invoking ninja)
Message: False
```

Because of this I've had to fall back to using `find_library`.

I'm using Meson version 0.39.0.

**Comment by QuLogic** [comment]
This might apply to many of the non-pkgconfig dependencies.

**Comment by nirbheek** [comment]
When we were fixing `PkgConfigDependency`, we didn't realize the same bugs applied to all other dependency types too. I have a WIP branch that fixes this issue and like a hundred other ones.

**Comment by bredelings** [comment]
@nirbheek What's the status on that branch?

**Comment by nirbheek** [comment]
That branch is lost to the sands of time, and the dependency code has been massively reorganised in the past year anyway, so it's probably best to implement it from scratch.

**Comment by bredelings** [comment]
So, should I try to hack in to the boost dependency specifically?  Is there a way to share version-checking code between other dependencies in dependencies/misc.py?

**Comment by nirbheek** [comment]
There is no built-in way yet, you would have to reorganize all the dependency types and factor out the version checking code to the parent class. If you don't want to go through all that effort, it's fine to add it just for Boost. That reorganising can happen later.

**Comment by jon-turney** [comment]
Closed by  #4035 (I think)

**Comment by dcbaker** [comment]
Agreed, closing. Please reopen if this is not solved by the mentioned PR.
